pkg/apis/pipeline/v1/merge.go

```diff
@@ -46,6 +46,11 @@ func MergeStepsWithStepTemplate(template *StepTemplate, steps []Step) ([]Step, e
 	}
 
 	for i, s := range steps {
+		// If the stepaction has not been fetched yet then do not merge.
```

Comment:
I am not sure I'm following here - why does if s.Ref is empty determine if the stepaction has been fetched?

Reply:
Because users can either provide a `ref` to `StepActions` (in which case it is not `nil`) or inline the actionable fields e.g. `image, command etc.` (in which case `ref` is `nil`). There is already a validation check that users cannot provide both `ref` and `image` (for example).

Reply:
And at webhook validation time, we have not yet done the remote resolution.

Reply:
At runtime, after we fetch the StepActions, we inline the content and set `ref` to nil explicitly. So when we validate at runtime, we can actually do this after the StepAction has been fetched. https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/taskrun/resources/taskspec.go#L145